### PR TITLE
Add Mean Average Precision validation method

### PR DESCRIPTION
**Pull request by Menooker**
## What changes were proposed in this pull request?

Add Mean Average Precision validation method for both classification and object detection.

## How was this patch tested?

UT and manually comparing results with other implementations of MAP

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

**Comment by zhangxiaoli73** [comment]
Add python wrapper for MAP

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala [hunk 1]
```diff
@@ -219,7 +219,325 @@ class Top1Accuracy[T: ClassTag](
 }
 
 /**
- * Caculate the percentage that target in output's top5 probability indexes
+ * Calculate the Mean Average Precision (MAP). The algorithm follows VOC Challenge after 2010
+ */
+class MeanAveragePrecision[T: ClassTag](k: Int, classes: Int)(
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala]: add parameter comments for k and classeses

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala [hunk 2]
```diff
@@ -219,7 +219,325 @@ class Top1Accuracy[T: ClassTag](
...
+  implicit ev: TensorNumeric[T]) extends ValidationMethod[T] {
+
+  require(classes > 0 && classes <= classes, s"The number of classes should be "
+    + s"> 0 and <= $classes, but got $classes")
+  require(k > 0, s"k should be > 0, but got $k")
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala]: what does K mean

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala [hunk 3]
```diff
@@ -219,7 +219,325 @@ class Top1Accuracy[T: ClassTag](
...
+        val precision = tp.toFloat / (j + 1)
+        val recall = tp.toFloat / posCnt(clz)
+        Iterator.single(recall, precision)
+      } else {
+        Iterator.empty
+      }
+    }
+
+    // get Average precision over each different recall
+    if (useVoc2007) {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala]: what does `useVoc2007` mean ?

Comment (Menooker) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala]: Pascal VOC uses different algorithms to calculate Average Precision before and after the year of 2007. It's a flag to indicate the algorithm.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala (hunk 2 above)
Comment (Menooker) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala]: K is the number of predictions (top-K confident items) that are calculated in AP. It is used in MAP of classification.  For object detection, it is set to -1 (count for all prediction)

Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala]: so please add class parameter comments

Comment (Menooker) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala]: comments updated

File: pyspark/bigdl/optim/optimizer.py [hunk 4]
```diff
@@ -68,6 +68,32 @@ class Top5Accuracy(JavaValue):
...
+        JavaValue.__init__(self, None, bigdl_type, k, classes)
+
+class MeanAveragePrecisionObjectDetection(JavaValue):
+    """
+    Caculate MeanAveragePrecision for Object Detection.
+
+    >>> MAPObj = MeanAveragePrecisionObjectDetection(20)
+    creating: createMeanAveragePrecisionObjectDetection
+    """
+    def __init__(self, classes, iou=0.5, useVoc2007=False, skipClass=-1, bigdl_type="float"):
```
Comment (wzhongyuan) [inline pyspark/bigdl/optim/optimizer.py]: following python stype, `use_voc`, `skip_class` ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala [hunk 5]
```diff
@@ -219,7 +219,327 @@ class Top1Accuracy[T: ClassTag](
...
+)
+  extends ValidationResult {
+
+  if (skipClass < 0) {
+    require(skipClass == -1, s"Invalid skipClass $skipClass")
+  } else {
+    require(skipClass >= 0 && skipClass < nClass, s"Invalid skipClass $skipClass")
+  }
+
+  def calculateClassAP(clz: Int): Float = {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala]: private ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala [hunk 6]
```diff
@@ -219,7 +219,327 @@ class Top1Accuracy[T: ClassTag](
...
+    val iymin = Math.max(ymin, y1)
+    val ixmax = Math.min(xmax, x2)
+    val iymax = Math.min(ymax, y2)
+    val inter = Math.max(ixmax - ixmin, 0) * Math.max(iymax - iymin, 0)
+    inter / ((x2 - x1) * (y2 - y1) + area - inter)
+  }
+}
+
+/** MeanAveragePrecision for Object Detection
+ * IMPORTANT: The labels in the target vector (Ground truth) begin from 0. BUT in the
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala]: begin with or start from

File: pyspark/bigdl/optim/optimizer.py [hunk 7]
```diff
@@ -68,6 +68,32 @@ class Top5Accuracy(JavaValue):
...
+    Caculate MeanAveragePrecision for Object Detection.
+
+    >>> MAPObj = MeanAveragePrecisionObjectDetection(20)
+    creating: createMeanAveragePrecisionObjectDetection
+    """
+    def __init__(self, classes, iou=0.5, use_voc2007=False, skip_class=-1, bigdl_type="float"):
+        """
+        :param classes: the number of classes
+        :param iou: the IOU threshold
+        :param useVoc2007: use validation method before voc2010 (i.e. voc2007)
```
Comment (wzhongyuan) [inline pyspark/bigdl/optim/optimizer.py]: please change the name in comments accordingly

**Comment by wzhongyuan** [comment]
LGTM,  Please run a Jenkins PASS before merging it.

**Comment by Menooker** [comment]
Jenkins pass: http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/845/

**Pull request merged**